### PR TITLE
Don't require TypeInfo in a minimal runtime if it's not needed

### PR DESCRIPTION
**Pull request by JinShil**
This is a followup to the following PRs

 * #7395 and #7768 for `ModuleInfo`
 * #7786 for `Throwable`

It is to `TypeInfo` what the above PRs were for `ModuleInfo` and `Throwable`

With this PR, it is now possible to have the same functionality as -betterC (without the -betterC compiler switch) by simply adding the following minimal runtime for executables...

```D
module object;

private alias extern(C) int function(char[][] args) MainFunc;
private extern (C) int _d_run_main(int argc, char** argv, MainFunc mainFunc)
{
    return mainFunc(null);  // assuming `void main()`
}
```

... or the following minimal runtime for libraries:

```D
module object;
```

For libraries it would be nice to not require an empty object.d file, but I'm leaving that for another potential PR.      UPDATE: See  #7825

Again, this will be of interest to users wishing to incrementally or partially port D to new platforms, users hoping to target resource constrained platforms, and users wishing to use D as a library from other languages without druntime.

### CODE REVIEW

File: src/dmd/typinf.d [hunk 1]
```diff
@@ -29,9 +29,16 @@ import dmd.visitor;
 extern (C++) void genTypeInfo(Type torig, Scope* sc)
 {
     //printf("Type::genTypeInfo() %p, %s\n", this, toChars());
+
+    if (!global.params.useTypeInfo)
```
Comment (JinShil) [inline src/dmd/typinf.d]: This block of code was modified to make it consistent with the following in statementsem.d

https://github.com/dlang/dmd/blob/ec4534ec63958b17167d636cd8b98824ce52f7d3/src/dmd/statementsem.d#L3626-L3636

File: src/dmd/typinf.d [hunk 2]
```diff
@@ -29,9 +29,16 @@ import dmd.visitor;
...
+    {
+        torig.error(Loc.initial, "`TypeInfo` cannot be used with -betterC");
+        fatal();
+    }
```
Comment (JinShil) [inline src/dmd/typinf.d]: This block of code was modified to make it consistent with the following in statementsem.d

https://github.com/dlang/dmd/blob/ec4534ec63958b17167d636cd8b98824ce52f7d3/src/dmd/statementsem.d#L3626-L3636

File: test/compilable/minimal.d [hunk 3]
```diff
@@ -2,7 +2,10 @@
 // POST_SCRIPT: compilable/extra-files/minimal/verify_symbols.sh
 // REQUIRED_ARGS: -c -defaultlib= runnable/extra-files/minimal/object.d
 
-// This test ensures an empty main built with a minimal runtime does not generate
-// ModuleInfo and exception handling code
+// This test ensures an empty main with a struct, built with a minimal runtime,
+// does not generate ModuleInfo or exception handling code, and does not
+// require TypeInfo
```
Comment (wilzbach) [inline test/compilable/minimal.d]: Don't you have to modify `verify_symbols` to check for `TypeInfo`?

Comment (JinShil) [reply test/compilable/minimal.d]: Yeah, best to be thorough.  Done.

**Comment by JinShil** [comment]
Ping @WalterBright, @ZombineDev, @ibuclaw... anyone, really.

**Review by jacob-carlborg — APPROVED** [review]

File: src/dmd/typinf.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/typinf.d]: Can we have a valid line for the error?

Comment (ibuclaw) [reply src/dmd/typinf.d]: Does this error apply for both explicit and implicit uses of RTTI?

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: src/dmd/typinf.d (hunk 2 above)
Comment (JinShil) [reply src/dmd/typinf.d]: > Does this error apply for both explicit and implicit uses of RTTI?

It applies to anything that ends up calling `genTypeInfo`, which, under the current implementation could be implicit or explicit uses of `TypeInfo`.  For example, trying to do a string concatenation will eventually call runtime hooks like `_d_arraycatnTX` which takes `TypeInfo` as one of its arguments.  See https://issues.dlang.org/show_bug.cgi?id=18312

IMO all of those runtime hooks should be changed to templates, that don't require `TypeInfo`, so we shouldn't need `TypeInfo` for string concatenation as long as the runtime hooks are implemented.  That's something I hope to champion after this PR (maybe 1 or 2 more) are merged.

TL;DR: Yes implicit and explicit uses of `TypeInfo` will trigger the error, but by replacing runtime hooks with templates (later) we may be able to eliminate such errors.

Comment (ibuclaw) [reply src/dmd/typinf.d]: Ok, but I'd still feel better if explicit uses got a file/line location to highlight where you're using RTTI.

Comment (JinShil) [reply src/dmd/typinf.d]: By "explicit uses" are you referring to `typeid`?  That's the only case I can think of where `TypeInfo` is explicitly used, and even then, it's indirect.  Please let me know if you have something else in mind.

Comment (rainers) [reply src/dmd/typinf.d]: I think showing the location of implicit uses is even more desirable, as you cannot grep for these as with explicit uses of `typeid`. That would mean to pass the location of an expression via semanticTypeInfo and getTypeInfoType.

That will probably also be helpful when generating type info more lazily as in https://github.com/dlang/dmd/pull/6561

Comment (JinShil) [reply src/dmd/typinf.d]: OK, I'll explore this.  Marking this as WIP while I try.

**Comment by JinShil** [comment]
I add a `Loc` parameter to `genTypeInfo` for reporting line numbers in the error messages.  This increased the diff, of course.  

I also add DDoc header comments to the function signatures I modified, and a couple more tests to increase coverage.

There is somewhat of a drawback to this, however.  For example, the following code will produce an error message about `TypeInfo` when it is not apparent, at all, why `TypeInfo` is needed.

```D
void test()
{
    string s = "abc"
    s = "(" ~ s ~ ")";  // main.d(8): Error: `object.TypeInfo` was not found
}
```

That is not actually a problem with this PR, but a general problem with D's ubiquitous use of runtime type information (i.e. `TypeInfo`) for things that are known at compile-time.  As I mentioned previously, I intend to champion an effort replace such uses of `TypeInfo` with templates, which should actually eliminate all such errors, and produce messages that are more meaningful.

cc @ibuclaw @rainers

File: src/dmd/e2ir.d [hunk 4]
```diff
@@ -949,7 +949,8 @@ Lagain:
                     r = (op == TOK.construct) ? RTLSYM_ARRAYSETCTOR : RTLSYM_ARRAYSETASSIGN;
                     evalue = el_una(OPaddr, TYnptr, evalue);
                     // This is a hack so we can call postblits on const/immutable objects.
-                    elem *eti = getTypeInfo(tb.unSharedOf().mutableOf(), irs);
+                    AggregateDeclaration ad = cast(AggregateDeclaration)tb;
+                    elem *eti = getTypeInfo(ad.loc, tb.unSharedOf().mutableOf(), irs);
```
Comment (rainers) [inline src/dmd/e2ir.d]: This location is the one of the aggregate declaration, not the using expression. For completeness, setArray should probably receive the expression to forward it here.

File: src/dmd/typinf.d [hunk 5]
```diff
@@ -24,14 +24,26 @@ import dmd.mtype;
...
+    if (!global.params.useTypeInfo)
+    {
+        torig.error(loc, "`TypeInfo` cannot be used with -betterC");
+        fatal();
+    }
+
     if (!Type.dtypeinfo)
     {
-        torig.error(Loc.initial, "TypeInfo not found. object.d may be incorrectly installed or corrupt, compile with -v switch");
+        torig.error(loc, "`object.TypeInfo` was not found");
```
Comment (rainers) [inline src/dmd/typinf.d]: Maybe add "but implicitly used" to the error message.

**Review by rainers — COMMENTED** [review]
Apart from nits LGTM

File: src/dmd/e2ir.d (hunk 4 above)
Comment (JinShil) [reply src/dmd/e2ir.d]: Done.  Thanks for the help!

File: src/dmd/typinf.d (hunk 5 above)
Comment (JinShil) [reply src/dmd/typinf.d]: Done.

**Review by rainers — APPROVED** [review]

**Comment by WalterBright** [comment]
This produced a regression https://issues.dlang.org/show_bug.cgi?id=18905
